### ISSUE TITLE
sql: add `date_part` builtin function

### ISSUE THREAD

**Issue by rafiss** [cockroachdb/cockroach#72437]
In Postgres, `date_part` is used as an almost-alias for the `extract` function. CockroachDB already has the `extract` builtin, but some tools call `date_part` instead. We should add it for compatibility.

See https://www.postgresql.org/docs/14/functions-datetime.html#FUNCTIONS-DATETIME-EXTRACT

Somewhat related to this issue: [Postgres 14 included this change](https://www.postgresql.org/docs/14/release-14.html) to `extract`:
> Change EXTRACT() to return type numeric instead of float8
> This avoids loss-of-precision issues in some usages. The old behavior can still be obtained by using the old underlying function date_part().
> Also, EXTRACT(date) now throws an error for units that are not part of the date data type.

Epic CRDB-1352

**Comment by otan** [comment]
this will be a PITA. namely, the backwards incompatible change will be the killer here.
we'll need to implement date_part for this to work.

**Comment by otan** [comment]
we also really shouldn't do this until we bump our support to postgres14

**Comment by rafiss** [comment]
I agree, no need to change our `extract` yet.

Here's more details on how PG handled the backwards-incompat: https://github.com/postgres/postgres/commit/a2da77cdb4661826482ebf2ddba1f953bc74afe4

> Changing the return type of an existing function is a
bit risky, so instead we do the following:  We implement a new set of
functions, which are now called "extract", in parallel to the existing
date_part functions.  They work the same way internally but use
numeric instead of float8.  The EXTRACT construct is now mapped by the
parser to these new extract functions.  That way, dumps of views
etc. from old versions (which would use date_part) continue to work
unchanged, but new uses will map to the new extract functions.
>
> Additionally, the reverse compilation of EXTRACT now reproduces the
original syntax, using the new mechanism introduced in
[40c24bf](https://github.com/postgres/postgres/commit/40c24bfef92530bd846e111c1742c2a54441c62c).

Filed https://github.com/cockroachdb/cockroach/issues/72523 so we can keep this issue just about adding date_part
